docs/configuration.md
# Configuration variables

## Vault configuration variables
* If you're deploying with **Vault**, set:
  * `VAULT_URL` - the vault url.
  * `VAULT_TOKEN` - the vault token is conceptually similar to a session cookie on a web site. Once a user authenticates, Vault returns a vault token which is used for future requests. The token is used by Vault to verify the identity of the client and to enforce the applicable ACL policies. This token is passed via HTTP headers.
  * `WALLET_NAME` - the wallet name that holds the authority account.
  * `AUTHORITY_ADDRESS` - this address is used to call the setAuthority interface on the plasma network.

* If there's **no Vault**, set:
  * `PRIVATE_KEY` - the private key. 

## Other configuration variables
- `ETHEREUM_RPC_URL` - the Ethereum RPC URL.
- `FINALITY_MARGIN` - the finality margin is number of Ethereum block confirmations to count before recognizing an event, by default is set to `10`.
- `ETHEREUM_NETWORK` - the Ethereum network.
- `TX_HASH_CONTRACT` - the contract deploy transaction hash.
- `AUTHORITY_ADDRESS` - the authority address, this address is used to call the setAuthority interface on the plasma network.
- `CONTRACT_ADDRESS_PLASMA_FRAMEWORK` - the plasma framework contract address that is deployed.
- `ETHEREUM_EVENTS_CHECK_INTERVAL_MS` -  the interval that checks for Ethereum events, by default is set to `8000`.
- `ETHEREUM_STALLED_SYNC_THRESHOLD_MS` - desynchronization threshold in milliseconds, by default is set to `20000`.
- `PREPARE_BLOCK_FOR_SUBMISSION_INTERVAL_MS` - the interval for preparing block for submission in milliseconds, by default is set to `10000`.
- `FEE_CLAIMER_ADDRESS` - the address that will claim the fee.
- `DATABASE_URL` - the postgres database url.
- `ENGINE_DB_POOL_SIZE` - the size of the engine database pool, by default is set to `10`.
- `ENGINE_DB_POOL_QUEUE_TARGET_MS` - the queue target of the engine database pool, by default is set to `100`.
- `ENGINE_DB_POOL_QUEUE_INTERVAL_MS` - the queue interval of the engine database pool, by default is set to `2000`
- `FEE_FEED_URL` - the fee feed url, by default is set to `http://localhost:4000/api/v1`
- `FEE_CHANGE_TOLERANCE_PERCENT` - by default is set to `25`
- `STORED_FEE_UPDATE_INTERVAL_MINUTES` - the interval to store fee updates in minutes, by default is set to `1`.
- `SENTRY_DSN` - Sentry automatically assigns you a Data Source Name (DSN) when you create a project to start monitoring events in your app. The DSN tells the SDK where to send the events. If this variable does not exist, the SDK will just not send any events.
- `HOSTNAME` - childchain's hostname.
- `APP_ENV` - app environment, needed for sentry.
- `DD_HOSTNAME` - the monitoring service hostname, by default is set to `datadog`.
- `DD_APM_PORT` - the datadog APM port, by default is set to `8126`.
- `BATCH_SIZE` - the batch size, by default is set to `10`
- `SYNC_THRESHOLD` - the sync threshold, by default is set to `100`
- `DD_DISABLED` - to disable or not to disable datadog, by default is set to `true`
- `RULES_FETCH_INTERVAL` - number of seconds between periodic retrieval of fee rules from GitHub, by default is set to `180`
- `PORT` - the childchain port, by default is set to `9656`

config/config.exs
import Config

parse_contracts = fn ->
  local_umbrella_path = Path.join([File.cwd!(), "../../", "localchain_contract_addresses.env"])

  contract_addreses_path =
    case File.exists?(local_umbrella_path) do
      true ->
        local_umbrella_path

      _ ->
        # CI/CD
        Path.join([File.cwd!(), "localchain_contract_addresses.env"])
    end

  contract_addreses_path
  |> File.read!()
  |> String.split("\n", trim: true)
  |> List.flatten()
  |> Enum.reduce(%{}, fn line, acc ->
    [key, value] = String.split(line, "=")
    Map.put(acc, key, value)
  end)
end

contracts = parse_contracts.()

config :engine,
  finality_margin: 10,
  network: "TEST",
  child_block_interval: 1000,
  fee_claimer_address: System.get_env("FEE_CLAIMER_ADDRESS") || "0x" <> Base.encode16(<<0::160>>),
  tx_hash_contract: contracts["TX_HASH_CONTRACT"],
  authority_address: contracts["AUTHORITY_ADDRESS"],
  plasma_framework: contracts["CONTRACT_ADDRESS_PLASMA_FRAMEWORK"],
  erc20_vault: contracts["CONTRACT_ADDRESS_ERC20_VAULT"],
  eth_vault: contracts["CONTRACT_ADDRESS_ETH_VAULT"],
  payment_exit_game: contracts["CONTRACT_ADDRESS_PAYMENT_EXIT_GAME"],
  block_submit_every_nth: 1

config :engine, Engine.Fee,
  fee_feed_url: "http://localhost:4000/api/v1/fees",
  fee_change_tolerance_percent: 25,
  stored_fee_update_interval_minutes: 1,
  fee_fetcher_check_interval_ms: 10_000,
  fee_buffer_duration_ms: 30_000

config :engine, Engine.Repo,
  backoff_type: :stop,
  # Have at most `:pool_size` DB connections on standby and serving DB queries.
  pool_size: String.to_integer(System.get_env("ENGINE_DB_POOL_SIZE") || "10"),
  # Wait at most `:queue_target` for a connection. If all connections checked out during
  # a `:queue_interval` takes more than `:queue_target`, then we double the `:queue_target`.
  # If checking out connections take longer than the new target, a DBConnection.ConnectionError is raised.
  # See: https://hexdocs.pm/db_connection/DBConnection.html#start_link/2-queue-config
  queue_target: String.to_integer(System.get_env("ENGINE_DB_POOL_QUEUE_TARGET_MS") || "100"),
  queue_interval: String.to_integer(System.get_env("ENGINE_DB_POOL_QUEUE_INTERVAL_MS") || "2000"),
  telemetry_prefix: [:engine, :repo]

config :engine, ecto_repos: [Engine.Repo]

config :ethereumex,
  http_options: [recv_timeout: 20_000]

config :logger, level: :info

config :logger, :console,
  format: "$date $time [$level] $metadata⋅$message⋅\n",
  discard_threshold: 2000,
  metadata: [:module, :function, :request_id, :trace_id, :span_id]

config :logger, Ink,
  name: "childchain",
  exclude_hostname: true

# THIS IS FOR APM - function traces
config :status, Status.Metric.Tracer,
  service: :backend,
  adapter: SpandexDatadog.Adapter,
  disabled?: true,
  type: :backend,
  env: "local-development-childchain-v2"

config :spandex, :decorators, tracer: Status.Metric.Tracer
config :spandex_phoenix, tracer: Status.Metric.Tracer

config :spandex_ecto, SpandexEcto.EctoLogger, tracer: Status.Metric.Tracer

# APMs are sent via HTTP requests
config :spandex_datadog,
  host: "localhost",
  port: 8126,
  batch_size: 10,
  sync_threshold: 100,
  http: HTTPoison

# Metrics are sent via UDP
config :statix,
  host: "localhost",
  port: 8125,
  tags: ["application:childchain-v2", "app_env:local-development"]

config :os_mon,
  disk_almost_full_threshold: 1,
  system_memory_high_watermark: 1

config :engine, Engine.Feefeed.Rules.Scheduler, interval: 180

config :ex_plasma,
  eip_712_domain: %{
    name: "OMG Network",
    salt: "0xfad5c7f626d80f9256ef01929f3beb96e058b8b4b0e3fe52d84f054c0e2a7a83",
    verifying_contract: System.get_env("CONTRACT_ADDRESS_PLASMA_FRAMEWORK"),
    version: "1"
  }

config :api,
  port: "PORT" |> System.get_env("9656") |> String.to_integer(),
  cors_enabled: true

import_config "#{Mix.env()}.exs"
